Core/Src/torque_manager.c

```diff
@@ -7,69 +7,111 @@
+            ref_torque_message.ref_torque[R_MOTOR] = ref_torque_lateral[R_MOTOR];
+            ref_torque_message.ref_torque[L_MOTOR] = ref_torque_lateral[L_MOTOR];
 
-			osMessageQueuePut(q_ref_torque_messageHandle, &ref_torque_message, 0, 0U);
+            osMessageQueuePut(q_ref_torque_messageHandle, &ref_torque_message, 0, 0U);
 
-			osDelay(RAMPA_DELAY);
+            osDelayUntil(tick);
 
-			break;
-		}
+            break;
 
-	}
+        default: ;// rampa de torque
+            uint32_t rampa_torque();
+            uint32_t ref_torque = rampa_torque();
 
-}
+            ref_torque_message.ref_torque[R_MOTOR] = ref_torque;
+            ref_torque_message.ref_torque[L_MOTOR] = ref_torque;
+
+            osMessageQueuePut(q_ref_torque_messageHandle, &ref_torque_message, 0, 0U);
+
+            osDelay(RAMPA_DELAY);
 
+            break;
+        }
 
+    }
+
+}
+
+// Rampa normal
 uint32_t rampa_torque() {
-	static uint32_t ref_torque_ant = 0;
-	uint32_t ref_torque = (modo_selecionado.torq_gain * throttle_percent) / 10;
-
-	if (ref_torque_ant > TORQUE_INIT_LIMITE) {					    // verifica se a referencia
-		if (ref_torque > ref_torque_ant + INC_TORQUE) {				// ja passou do ponto de inflexao
-			ref_torque = ref_torque_ant + INC_TORQUE;				// e aplica o incremento mais agressivo
-		}
-	} else {
-		if (ref_torque > ref_torque_ant + INC_TORQUE_INIT) {		// caso contrario usa o
-			ref_torque = ref_torque_ant + INC_TORQUE_INIT;			// incremento da primeira rampa
-		}
-	}
-	ref_torque_ant = ref_torque;									// aplica a referencia calculada
-
-	return ref_torque;
+    static uint32_t ref_torque_ant = 0;
+    uint32_t ref_torque = (uint32_t)((float)(modo_selecionado.torq_gain * throttle_percent) / 10);
+
+    if (ref_torque_ant > TORQUE_INIT_LIMITE) {					    // verifica se a referencia
+        if (ref_torque > ref_torque_ant + INC_TORQUE) {				// ja passou do ponto de inflexao
+            ref_torque = ref_torque_ant + INC_TORQUE;				// e aplica o incremento mais agressivo
+        }
+    } else {
+        if (ref_torque > ref_torque_ant + INC_TORQUE_INIT) {		// caso contrario usa o
+            ref_torque = ref_torque_ant + INC_TORQUE_INIT;			// incremento da primeira rampa
+        }
+    }
+    ref_torque_ant = ref_torque;									// aplica a referencia calculada
+
+    return ref_torque;
+}
+
+// Rampa vetorização
+void rampa_torque_lateral(lateral_t *ref_torque_decrease, uint32_t *ref_torque) {
+    static uint32_t ref_torque_ant[2] = {0, 0};
+    ref_torque[R_MOTOR] = (uint32_t)((float)(modo_selecionado.torq_gain * throttle_percent) / 10);
+    ref_torque[L_MOTOR] = (uint32_t)((float)(modo_selecionado.torq_gain * throttle_percent) / 10);
+
+    if (ref_torque_decrease->ref_wheel == R_MOTOR)
+        ref_torque[R_MOTOR] = (uint32_t)((float)ref_torque[R_MOTOR] - ref_torque_decrease->ref_decrease);
+    else
+        ref_torque[L_MOTOR] = (uint32_t)((float)ref_torque[L_MOTOR] - ref_torque_decrease->ref_decrease);
```

Comment:
Se `ref_torque_decrease->ref_decrease` for maior que `ref_torque[X_MOTOR]` acho que pode dá problema, pq como é unsigned ele vai acabar dando um overflow (n sei se essa é a expressão certa) e o valor de `ref_torque` vai ser gigante

Reply:
verdade. Recomendo vc usar uma lógica tipo
```cpp
ref_torque[R_MOTOR] = (uint32_t) max((int32_t)ref_torque[R_MOTOR] - ref_torque_decrease->ref_decrease, 0);
```
basicamente ele vai pegar o que é maior: ou essa subtração ou 0, então se der underflow ele trava em 0.
Nesse caso n precisa do float, mas vc vai precisar tratar essa substração como signed pra lógica do max funcionar.

E outra coisa: diferente de C++, acho que C n fornece uma função nativa de max e min. Recomendo vc criar uma lá nos arquivos de util e usar ela aqui. Dps olha na internet como criar essas funções genéricas. Pra ser utilizável pra vários tipos, precisa ser uma macro, parecido com aquilo q eu fiz nas queues de variáveis globais